File: apps/api/app/services/quality/metrics.py
"""Response Quality Checker — computes every Trust Panel metric.

P0 honesty statement: these are heuristic estimators, not learned evaluators —
but they run on real outputs and genuinely move when output quality degrades
(tests/test_quality.py proves collapse detection on cloned reactions).
The evaluation roadmap (docs/evaluation-framework.md) upgrades each estimator.

Metric map (spec section 9):
1. persona_adherence      — trait->behavior rank correlations across the swarm
2. product_grounding      — do reactions reference actual stimulus content?
3. generic_response_rate  — vague filler detection
4. duplicate_objection_rate — near-identical objection share
5. objection_entropy      — diversity of the objection distribution
6. segment_variance       — do segments actually respond differently?
7. collapse_risk          — combined mode-collapse indicator
8. benchmark_confidence   — do we have reference data for this category?
"""

from __future__ import annotations

import json
from collections import Counter
from pathlib import Path

from ...schemas.persona import Persona
from ...schemas.quality import QualityMetrics
from ...schemas.reaction import PersonaReaction
from ...utils.text import (
    clamp,
    normalize_objection,
    rank_correlation,
    shannon_entropy_norm,
    stddev,
)
from ..stimulus_parser import StimulusFeatures
from .consistency_checker import criteria_consistency_score

GENERIC_PHRASES = [
    "seems useful", "sounds useful", "interesting product", "sounds good",
    "i would consider it", "this product is innovative", "nice product",
    "cool idea", "looks good", "pretty good product", "might be useful",
]


def compute_quality(
    personas: list[Persona],
    reactions: list[PersonaReaction],
    features: StimulusFeatures,
    benchmark_dir: Path | None = None,
) -> QualityMetrics:
    notes: list[str] = []
    n = max(1, len(reactions))
    by_id = {p.persona_id: p for p in personas}
    paired = [(by_id[r.persona_id], r) for r in reactions if r.persona_id in by_id]

    adherence = _persona_adherence(paired, features, notes)
    grounding = _product_grounding(reactions, features)
    generic_rate = _generic_rate(reactions, features)
    entropy_norm = _objection_entropy(reactions)
    dup_rate = _duplicate_rate(reactions)
    seg_var = _segment_variance(reactions)
    age_var = _age_cohort_variance(reactions)
    consistency = criteria_consistency_score(personas, reactions)

    # ---- collapse risk: verbatim duplication + low theme entropy + likelihood
    # concentration. NOTE the deliberate distinction: many personas raising the
    # SAME THEME ("no pricing") is *signal*, not collapse — themes live in
    # entropy. Collapse is when full outputs are near-verbatim copies, so the
    # duplicate metric runs on whole quotes, not objection themes.
    concentration = _likelihood_concentration(reactions)
    collapse_score = clamp(0.35 * dup_rate + 0.35 * (1.0 - entropy_norm) + 0.30 * concentration)
    collapse_level = "low" if collapse_score < 0.33 else "medium" if collapse_score < 0.60 else "high"
    if collapse_level != "low":
        notes.append(
            f"Collapse signal: {dup_rate:.0%} duplicate objections, "
            f"entropy {entropy_norm:.2f}, likelihood concentration {concentration:.2f}."
        )

    entropy_level = "low" if entropy_norm < 0.45 else "medium" if entropy_norm < 0.70 else "high"
    var_level = _variance_strength(seg_var)
    age_var_level = _variance_strength(age_var)

    bench_level, bench_cat = _benchmark_confidence(features, benchmark_dir, notes)

    if consistency < 0.6:
        notes.append(
            f"Internal consistency low: only {consistency:.0%} of personas pass all "
            "consistency rules (trust/buy, price/WTP, proof/trust, uniform criteria) — "
            "many reactions contradict themselves."
        )

    notes.append(
        f"Metrics computed over {n} reactions across "
        f"{len({r.segment for r in reactions})} segments; heuristic P0 estimators."
    )

    return QualityMetrics(
        persona_adherence=round(adherence, 3),
        product_grounding=round(grounding, 3),
        generic_response_rate=round(generic_rate, 3),
        duplicate_objection_rate=round(dup_rate, 3),
        objection_entropy=entropy_level,
        objection_entropy_score=round(entropy_norm, 3),
        segment_variance=var_level,
        segment_variance_score=round(seg_var, 4),
        age_cohort_variance=age_var_level,
        criteria_consistency=round(consistency, 3),
        collapse_risk=collapse_level,
        collapse_risk_score=round(collapse_score, 3),
        benchmark_confidence=bench_level,
        benchmark_category=bench_cat,
        notes=notes,
    )


# ------------------------------------------------------------------- components

def _map_corr(corr: float, expected_sign: int) -> float:
    """Map a rank correlation to [0,1]: 0.5 = no signal, 1.0 = strong signal in
    the expected direction, 0.0 = strong signal in the WRONG direction."""
    return clamp(0.5 + expected_sign * corr * 1.4)


def _persona_adherence(paired: list[tuple[Persona, PersonaReaction]],
                       features: StimulusFeatures, notes: list[str]) -> float:
    if len(paired) < 10:
        return 0.5
    ps = [p.price_sensitivity for p, _ in paired]
    # normalize max_price by budget to remove the income confound
    price_ratio = [r.max_price / max(1.0, p.monthly_budget_usd) for p, r in paired]
    skept = [p.skepticism for p, _ in paired]
    trust = [p.brand_trust for p, _ in paired]
    likelihood = [r.buy_likelihood for _, r in paired]

    components = [
        _map_corr(rank_correlation(ps, price_ratio), expected_sign=-1),   # sensitive -> lower WTP
        _map_corr(rank_correlation(skept, likelihood), expected_sign=-1),  # skeptics -> lower intent
        _map_corr(rank_correlation(trust, likelihood), expected_sign=+1),  # trusting -> higher intent
    ]
    if features.mentions_ai:
        novelty = [p.novelty_seeking for p, _ in paired]
        components.append(_map_corr(rank_correlation(novelty, likelihood), expected_sign=+1))

    score = sum(components) / len(components)
    if score < 0.55:
        notes.append("Persona adherence weak: trait-behavior correlations below target.")
    return score


def _mentions_stimulus(text: str, features: StimulusFeatures) -> bool:
    low = text.lower()
    if any(a in low for a in features.anchor_set()):
        return True
    # price references count as grounding when the stimulus actually has prices
    if features.has_pricing and "$" in text:
        return True
    return False


def _product_grounding(reactions: list[PersonaReaction],
                       features: StimulusFeatures) -> float:
    """Approximate: share of reactions whose text references stimulus anchors or
    detected prices. Feature-level objections ('no pricing shown') ground in
    *absence*, which token matching can't see — so treat this as a lower bound."""
    if not reactions:
        return 0.0
    hits = sum(
        1 for r in reactions
        if _mentions_stimulus(f"{r.first_objection} {r.quote} {r.positive_trigger}", features)
    )
    return hits / len(reactions)


def _generic_rate(reactions: list[PersonaReaction],
                  features: StimulusFeatures) -> float:
    if not reactions:
        return 1.0

    def is_generic(r: PersonaReaction) -> bool:
        text = f"{r.first_objection} {r.quote}".lower()
        if not r.first_objection.strip():
            return True
        has_phrase = any(ph in text for ph in GENERIC_PHRASES)
        return has_phrase and not _mentions_stimulus(text, features)

    return sum(1 for r in reactions if is_generic(r)) / len(reactions)


def _objection_entropy(reactions: list[PersonaReaction]) -> float:
    """Diversity of objection THEMES (normalized text). High = varied concerns;
    low = the swarm converged on one concern (fine) or collapsed (check dup)."""
    keys = [normalize_objection(r.first_objection) for r in reactions if r.first_objection]
    if not keys:
        return 0.0
    return shannon_entropy_norm(list(Counter(keys).values()))


def _duplicate_rate(reactions: list[PersonaReaction], sample_size: int = 300) -> float:
    """Near-verbatim duplication of FULL outputs (normalized quote text).
    A healthy swarm repeats themes but not sentences; a collapsed model repeats
    sentences. Computed on a fixed-size seeded subsample so the metric is
    comparable across swarm sizes (raw distinct/n inevitably falls as n grows).
    0 = all quotes distinct, ->1 = clones."""
    import random as _random

    keys = [normalize_objection(r.quote) for r in reactions if r.quote.strip()]
    if not keys:
        return 1.0
    if len(keys) > sample_size:
        keys = _random.Random(0).sample(keys, sample_size)
    return 1.0 - len(set(keys)) / len(keys)


def _variance_strength(var: float) -> str:
    """Shared weak/moderate/strong mapping for any stddev-of-group-means
    variance score (segment_variance and age_cohort_variance both use it)."""
    return "weak" if var < 0.03 else "moderate" if var < 0.07 else "strong"


def _grouped_mean_variance(reactions: list[PersonaReaction], key) -> float:
    """stddev of per-group mean buy_likelihood, grouped by `key(reaction)`."""
    by_group: dict[str, list[float]] = {}
    for r in reactions:
        by_group.setdefault(key(r) or "unknown", []).append(r.buy_likelihood)
    means = [sum(v) / len(v) for v in by_group.values() if v]
    return stddev(means) if len(means) >= 2 else 0.0


def _segment_variance(reactions: list[PersonaReaction]) -> float:
    return _grouped_mean_variance(reactions, key=lambda r: r.segment)


def _age_cohort_variance(reactions: list[PersonaReaction]) -> float:
    """Do life-stage cohorts (teen_student, early_career, parent_family, ...)
    actually respond differently? Same stddev-of-means shape as
    _segment_variance, grouped by life_stage instead of segment."""
    return _grouped_mean_variance(reactions, key=lambda r: r.life_stage)


def _likelihood_concentration(reactions: list[PersonaReaction]) -> float:
    """Max histogram-bin share of buy_likelihood, rescaled so a uniform spread
    scores ~0 and 'everyone in one band' scores ~1 (mode-collapse smell)."""
    if not reactions:
        return 1.0
    bins = [0] * 10
    for r in reactions:
        bins[min(9, int(r.buy_likelihood * 10))] += 1
    max_share = max(bins) / len(reactions)
    return clamp((max_share - 0.15) / 0.85)


def _benchmark_confidence(features: StimulusFeatures, benchmark_dir: Path | None,
                          notes: list[str]) -> tuple[str, str | None]:
    """P0: do we have sample benchmark data for the detected category?
    Deliberately capped at 'medium' until real calibration studies land
    (never overstate trust — that's the whole point of the panel)."""
    if benchmark_dir is None or not benchmark_dir.exists():
        return "low", features.category
    index_file = benchmark_dir / "index.json"
    if not index_file.exists():
        return "low", features.category
    try:
        index = json.loads(index_file.read_text())
    except (OSError, json.JSONDecodeError):
        return "low", features.category
    entry = index.get(features.category)
    if not entry:
        notes.append(
            f"No benchmark samples for category '{features.category}' — "
            "treat absolute numbers as directional only."
        )
        return "low", features.category
    notes.append(
        f"Benchmark samples found for '{features.category}' "
        f"({entry.get('samples', 0)} reference sets) — confidence capped at medium for P0."
    )
    return "medium", features.category

File: apps/web/lib/server/engine/quality/metrics.ts
/**
 * Response Quality Checker — computes every Trust Panel metric. Port of
 * apps/api/app/services/quality/metrics.py.
 *
 * Benchmark note: no benchmark sample corpus ships in the serverless runtime,
 * so benchmark_confidence is always "low" (honest — never overstate trust).
 * This matches the Python behavior when benchmark_dir is absent.
 */

import {
  clamp,
  normalizeObjection,
  rankCorrelation,
  round,
  shannonEntropyNorm,
  stddev,
} from "../text";
import type { StimulusFeatures } from "../stimulusParser";
import { anchorSet } from "../stimulusParser";
import type { Persona, PersonaReaction } from "../types";
import type { Level, QualityMetrics, Strength } from "../report";
import { criteriaConsistencyScore } from "./consistency";

const GENERIC_PHRASES = [
  "seems useful", "sounds useful", "interesting product", "sounds good",
  "i would consider it", "this product is innovative", "nice product",
  "cool idea", "looks good", "pretty good product", "might be useful",
];

export function computeQuality(
  personas: Persona[],
  reactions: PersonaReaction[],
  features: StimulusFeatures,
): QualityMetrics {
  const notes: string[] = [];
  const n = Math.max(1, reactions.length);
  const byId = new Map(personas.map((p) => [p.persona_id, p]));
  const paired = reactions.filter((r) => byId.has(r.persona_id)).map((r) => [byId.get(r.persona_id)!, r] as const);

  const adherence = personaAdherence(paired, features, notes);
  const grounding = productGrounding(reactions, features);
  const genericRate = genericRateOf(reactions, features);
  const entropyNorm = objectionEntropy(reactions);
  const dupRate = duplicateRate(reactions);
  const segVar = segmentVariance(reactions);
  const ageVar = ageCohortVariance(reactions);
  const consistency = criteriaConsistencyScore(personas, reactions);

  const concentration = likelihoodConcentration(reactions);
  const collapseScore = clamp(0.35 * dupRate + 0.35 * (1.0 - entropyNorm) + 0.3 * concentration);
  const collapseLevel: Level = collapseScore < 0.33 ? "low" : collapseScore < 0.6 ? "medium" : "high";
  if (collapseLevel !== "low") {
    notes.push(
      `Collapse signal: ${pct(dupRate)} duplicate objections, entropy ${entropyNorm.toFixed(2)}, likelihood concentration ${concentration.toFixed(2)}.`,
    );
  }

  const entropyLevel: Level = entropyNorm < 0.45 ? "low" : entropyNorm < 0.7 ? "medium" : "high";
  const varLevel = varianceStrength(segVar);
  const ageVarLevel = varianceStrength(ageVar);

  const benchLevel: Level = "low";
  const benchCat = features.category;
  notes.push(`No benchmark corpus in the serverless runtime — treat absolute numbers for '${benchCat}' as directional only.`);

  if (consistency < 0.6) {
    notes.push(
      `Internal consistency low: only ${pct(consistency)} of personas pass all consistency rules (trust/buy, price/WTP, proof/trust, uniform criteria) — many reactions contradict themselves.`,
    );
  }

  const segCount = new Set(reactions.map((r) => r.segment)).size;
  notes.push(`Metrics computed over ${n} reactions across ${segCount} segments; heuristic P0 estimators.`);

  return {
    persona_adherence: round(adherence, 3),
    product_grounding: round(grounding, 3),
    generic_response_rate: round(genericRate, 3),
    duplicate_objection_rate: round(dupRate, 3),
    objection_entropy: entropyLevel,
    objection_entropy_score: round(entropyNorm, 3),
    segment_variance: varLevel,
    segment_variance_score: round(segVar, 4),
    age_cohort_variance: ageVarLevel,
    criteria_consistency: round(consistency, 3),
    collapse_risk: collapseLevel,
    collapse_risk_score: round(collapseScore, 3),
    benchmark_confidence: benchLevel,
    benchmark_category: benchCat,
    notes,
  };
}

function pct(v: number): string {
  return `${Math.round(v * 100)}%`;
}

function mapCorr(corr: number, expectedSign: number): number {
  return clamp(0.5 + expectedSign * corr * 1.4);
}

function personaAdherence(
  paired: readonly (readonly [Persona, PersonaReaction])[],
  features: StimulusFeatures,
  notes: string[],
): number {
  if (paired.length < 10) return 0.5;
  const ps = paired.map(([p]) => p.price_sensitivity);
  const priceRatio = paired.map(([p, r]) => r.max_price / Math.max(1.0, p.monthly_budget_usd));
  const skept = paired.map(([p]) => p.skepticism);
  const trust = paired.map(([p]) => p.brand_trust);
  const likelihood = paired.map(([, r]) => r.buy_likelihood);

  const components = [
    mapCorr(rankCorrelation(ps, priceRatio), -1),
    mapCorr(rankCorrelation(skept, likelihood), -1),
    mapCorr(rankCorrelation(trust, likelihood), 1),
  ];
  if (features.mentionsAi) {
    const novelty = paired.map(([p]) => p.novelty_seeking);
    components.push(mapCorr(rankCorrelation(novelty, likelihood), 1));
  }
  const score = components.reduce((s, v) => s + v, 0) / components.length;
  if (score < 0.55) notes.push("Persona adherence weak: trait-behavior correlations below target.");
  return score;
}

function mentionsStimulus(text: string, features: StimulusFeatures): boolean {
  const low = text.toLowerCase();
  for (const a of anchorSet(features)) if (low.includes(a)) return true;
  if (features.hasPricing && text.includes("$")) return true;
  return false;
}

function productGrounding(reactions: PersonaReaction[], features: StimulusFeatures): number {
  if (reactions.length === 0) return 0.0;
  const hits = reactions.filter((r) =>
    mentionsStimulus(`${r.first_objection} ${r.quote} ${r.positive_trigger}`, features),
  ).length;
  return hits / reactions.length;
}

function genericRateOf(reactions: PersonaReaction[], features: StimulusFeatures): number {
  if (reactions.length === 0) return 1.0;
  const isGeneric = (r: PersonaReaction): boolean => {
    const text = `${r.first_objection} ${r.quote}`.toLowerCase();
    if (!r.first_objection.trim()) return true;
    const hasPhrase = GENERIC_PHRASES.some((ph) => text.includes(ph));
    return hasPhrase && !mentionsStimulus(text, features);
  };
  return reactions.filter(isGeneric).length / reactions.length;
}

function objectionEntropy(reactions: PersonaReaction[]): number {
  const keys = reactions.filter((r) => r.first_objection).map((r) => normalizeObjection(r.first_objection));
  if (keys.length === 0) return 0.0;
  const counts = new Map<string, number>();
  for (const k of keys) counts.set(k, (counts.get(k) ?? 0) + 1);
  return shannonEntropyNorm(Array.from(counts.values()));
}

function duplicateRate(reactions: PersonaReaction[], sampleSize = 300): number {
  let keys = reactions.filter((r) => r.quote.trim()).map((r) => normalizeObjection(r.quote));
  if (keys.length === 0) return 1.0;
  if (keys.length > sampleSize) {
    // Deterministic subsample (seeded, mirrors random.Random(0).sample intent).
    keys = seededSample(keys, sampleSize);
  }
  return 1.0 - new Set(keys).size / keys.length;
}

function seededSample<T>(arr: T[], k: number): T[] {
  // Deterministic partial Fisher–Yates with a fixed seed (0).
  let state = 0x9e3779b9 >>> 0;
  const rand = (): number => {
    state = (state + 0x6d2b79f5) >>> 0;
    let t = state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
  const pool = arr.slice();
  const n = pool.length;
  for (let i = 0; i < k; i++) {
    const j = i + Math.floor(rand() * (n - i));
    const tmp = pool[i];
    pool[i] = pool[j];
    pool[j] = tmp;
  }
  return pool.slice(0, k);
}

function varianceStrength(v: number): Strength {
  return v < 0.03 ? "weak" : v < 0.07 ? "moderate" : "strong";
}

function groupedMeanVariance(reactions: PersonaReaction[], key: (r: PersonaReaction) => string): number {
  const byGroup = new Map<string, number[]>();
  for (const r of reactions) {
    const g = key(r) || "unknown";
    if (!byGroup.has(g)) byGroup.set(g, []);
    byGroup.get(g)!.push(r.buy_likelihood);
  }
  const means = Array.from(byGroup.values()).filter((v) => v.length > 0).map((v) => v.reduce((s, x) => s + x, 0) / v.length);
  return means.length >= 2 ? stddev(means) : 0.0;
}

function segmentVariance(reactions: PersonaReaction[]): number {
  return groupedMeanVariance(reactions, (r) => r.segment);
}

function ageCohortVariance(reactions: PersonaReaction[]): number {
  return groupedMeanVariance(reactions, (r) => r.life_stage);
}

function likelihoodConcentration(reactions: PersonaReaction[]): number {
  if (reactions.length === 0) return 1.0;
  const bins = new Array<number>(10).fill(0);
  for (const r of reactions) bins[Math.min(9, Math.floor(r.buy_likelihood * 10))] += 1;
  const maxShare = Math.max(...bins) / reactions.length;
  return clamp((maxShare - 0.15) / 0.85);
}
